solo-io/gloo
# Istio spiffe volumes

# Description

Add volumes, mounted volumes, and env vars to GatewayProxy template as required to support newer versions of Istio

Also update the default Helm value for Istio image to be one we claim to support

## Helm changes
- Add `CANONICAL_SERVICE` and `CANONICAL_REVISION` env vars to GatewayProxy IstioProxy container, from corresponding labels
- Add `workload-socket` and `workload-secret` volumes to GatewayProxy and mount them at `/var/run/secrets/workload-spiffe-uds` and `/var/run/secrets/workload-spiffe-credentials` respectively in the IstioProxy container
- Update default `global.glooMtls.istioProxy.image.tag` value to `1.14.6`, as the the previous was no longer supported


# Context

These changes match exactly what is requested in the issue and what is required by newer versions of Istio

I was not able to find explicit documentation of what versions introduced the requirement to have the spiffe volumes, or even what they are...the best I found was [this page](https://istio.io/latest/docs/ops/integrations/spire/) which references one of the volumes
Instead I went off the example config given in the issue, which seems to be accurate

I do not believe additional documentation is necessary, as this is effectively a fix for supporting Istio versions we already claim to support and does not require any change in user behavior

This will be backported to 1.15, 1.14, and 1.13

## Interesting decisions
 
As previously stated, it's unclear exactly what Istio version starts to require the spiffe volumes, however the presence of the (initially empty) volumes should not break usage of Istio versions that did not require them

The earliest version of Istio we claim to support on any LTS version of Gloo is 1.11, so it made sense to bump the default value to a version we support
I chose 1.14.6 as it's the version specifically referenced in the issue and supported by all LTS versions, however I'm happy to change to a different value if we prefer to go with the minimum or maximum per LTS version

I added to Helm unit tests to test for newly added values, without adding tests for previously supported env vars, volumes, or mounts
It's preferable to have the strongest tests possible for new functionality, but seems out of scope to exhaustively test every aspect of existing functionality

## Testing steps

- Create a fresh KinD cluster and install Istio 1.14.6
- Installing Gloo from `main` with the following values set (in addition to the default values for everything else):
```
global:
  istioIntegration:
    whitelistDiscovery: true
    enableIstioSidecarOnGateway: true
  istioSDS:
    enabled: true
```
- Observe the expected error in the GatewayProxy `istio-proxy` container
- Helm uninstall `gloo` deployment, modify the template/checkout the `istio-spiffe-volumes` branch and Helm install from the local chart
- Observe the GatewayProxy pod comes up healthily

# Checklist:
